// @arkecosystem/core-magistrate-crypto/dist/enums.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var MagistrateTransactionType;
(function (MagistrateTransactionType) {
    MagistrateTransactionType[MagistrateTransactionType["BusinessRegistration"] = 0] = "BusinessRegistration";
    MagistrateTransactionType[MagistrateTransactionType["BusinessResignation"] = 1] = "BusinessResignation";
    MagistrateTransactionType[MagistrateTransactionType["BusinessUpdate"] = 2] = "BusinessUpdate";
    MagistrateTransactionType[MagistrateTransactionType["BridgechainRegistration"] = 3] = "BridgechainRegistration";
    MagistrateTransactionType[MagistrateTransactionType["BridgechainResignation"] = 4] = "BridgechainResignation";
    MagistrateTransactionType[MagistrateTransactionType["BridgechainUpdate"] = 5] = "BridgechainUpdate";
    MagistrateTransactionType[MagistrateTransactionType["Entity"] = 6] = "Entity";
})(MagistrateTransactionType = exports.MagistrateTransactionType || (exports.MagistrateTransactionType = {}));
exports.MagistrateTransactionGroup = 2;
var MagistrateTransactionStaticFees;
(function (MagistrateTransactionStaticFees) {
    MagistrateTransactionStaticFees["BusinessRegistration"] = "5000000000";
    MagistrateTransactionStaticFees["BusinessResignation"] = "5000000000";
    MagistrateTransactionStaticFees["BusinessUpdate"] = "5000000000";
    MagistrateTransactionStaticFees["BridgechainRegistration"] = "5000000000";
    MagistrateTransactionStaticFees["BridgechainResignation"] = "5000000000";
    MagistrateTransactionStaticFees["BridgechainUpdate"] = "5000000000";
    MagistrateTransactionStaticFees["EntityRegister"] = "5000000000";
    MagistrateTransactionStaticFees["EntityUpdate"] = "500000000";
    MagistrateTransactionStaticFees["EntityResign"] = "500000000";
})(MagistrateTransactionStaticFees = exports.MagistrateTransactionStaticFees || (exports.MagistrateTransactionStaticFees = {}));
// Entity types can be any integer between 0-255 but this enum keeps track of what has been assigned already
var EntityType;
(function (EntityType) {
    EntityType[EntityType["Business"] = 0] = "Business";
    EntityType[EntityType["Bridgechain"] = 1] = "Bridgechain";
    EntityType[EntityType["Developer"] = 2] = "Developer";
    EntityType[EntityType["Plugin"] = 3] = "Plugin";
    EntityType[EntityType["Delegate"] = 4] = "Delegate";
})(EntityType = exports.EntityType || (exports.EntityType = {}));
var EntityAction;
(function (EntityAction) {
    EntityAction[EntityAction["Register"] = 0] = "Register";
    EntityAction[EntityAction["Update"] = 1] = "Update";
    EntityAction[EntityAction["Resign"] = 2] = "Resign";
})(EntityAction = exports.EntityAction || (exports.EntityAction = {}));
